sentry-core/src/main/java/io/sentry/core/cache/DiskCache.java

```diff
@@ -0,0 +1,152 @@
+import static io.sentry.core.SentryLevel.DEBUG;
+import static io.sentry.core.SentryLevel.ERROR;
+import static io.sentry.core.SentryLevel.WARNING;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import io.sentry.core.ISerializer;
+import io.sentry.core.SentryEvent;
+import io.sentry.core.SentryLevel;
+import io.sentry.core.SentryOptions;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A simple cache implementation storing the events to a disk, each event in a separater file in a configured directory.
+ */
+public final class DiskCache implements IEventCache {
+  /**
+   * File suffix added to all serialized event files.
+   */
+  public static final String FILE_SUFFIX = ".sentry-event";
+
+  @SuppressWarnings("CharsetObjectCanBeUsed")
+  private static final Charset UTF8 = Charset.forName("UTF-8");
+
+  private final File directory;
+  private final int maxSize;
+  private final ISerializer serializer;
+  private final SentryOptions options;
+
+  public DiskCache(SentryOptions options) {
+    this.directory = new File(options.getCacheDirPath());
+    this.maxSize = options.getCacheDirSize();
+    this.serializer = options.getSerializer();
+    this.options = options;
+
+    try {
+      //noinspection ResultOfMethodCallIgnored
+      directory.mkdirs();
+      checkDirectoryValid();
+    } catch (IllegalStateException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new IllegalStateException("Failed to initialize the directory for caching Sentry events.", e);
+    }
+  }
+
+  @Override
+  public void store(SentryEvent event) {
+    if (getNumberOfStoredEvents() >= maxSize) {
+      logIfNotNull(options.getLogger(), SentryLevel.WARNING, "Disk cache full. Not storing event {}", event);
+      return;
+    }
+
+    File eventFile = getEventFile(event);
+    if (eventFile.exists()) {
+      logIfNotNull(options.getLogger(), DEBUG,
+        "Not adding Event to offline storage because it already exists: %s", eventFile.getAbsolutePath());
+      return;
+    } else {
+      logIfNotNull(options.getLogger(), DEBUG, "Adding Event to offline storage: %s", eventFile.getAbsolutePath());
+    }
+
+    try (FileOutputStream fileOutputStream = new FileOutputStream(eventFile);
+         Writer wrt = new OutputStreamWriter(fileOutputStream, UTF8)) {
+      serializer.serialize(event, wrt);
+    } catch (Exception e) {
+      logIfNotNull(options.getLogger(), ERROR, "Error writing Event to offline storage: %s", event.getEventId());
+    }
+  }
+
+  @Override
+  public void discard(SentryEvent event) {
+    File eventFile = getEventFile(event);
+        if (eventFile.exists()) {
```

Comment:
`exists` followed by `delete` seems redundant or racey. Is there a way to get`delete` to returns whether the file was there and deleted or didn't exist?

Reply:
Well, the check is there merely for correct reporting to the log. We handle the failure to delete by checking the return value of `eventFile.delete()` in the code below, but unfortunately the File API doesn't provide the reason for the failure. `java.nio.files.Files#delete` is better in that regard because it throws an exception, but I assume it is not supported on our minimum Android API level.